File: backend/routers/tools.py
"""Standalone tool endpoints: web search and RAG query."""

from typing import List

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from dependencies import get_rag_service, get_search_service
from services.rag_service import RAGService
from services.search_service import SearchService

router = APIRouter()


class SearchRequest(BaseModel):
    query: str = Field(..., min_length=1)
    max_results: int = 5
    scrape_content: bool = True


class RAGQueryRequest(BaseModel):
    query: str = Field(..., min_length=1)
    file_ids: List[str] = []
    top_k: int = 3


@router.post("/search")
async def web_search(
    request: SearchRequest, search: SearchService = Depends(get_search_service)
):
    """Web search with optional content scraping."""
    results = await search.search(
        request.query, request.max_results, request.scrape_content
    )
    return {"results": results}


@router.post("/rag/query")
async def rag_query(
    request: RAGQueryRequest, rag: RAGService = Depends(get_rag_service)
):
    """Query the RAG vector store."""
    results = await rag.query(
        query=request.query, file_ids=request.file_ids, top_k=request.top_k
    )
    return {"results": results}

File: src/api/tools.ts
import axios from 'axios'
import type { SearchResult, RAGResult } from '../types'

const API_BASE_URL = '/api'

export async function webSearch(query: string, maxResults = 5): Promise<SearchResult[]> {
  try {
    const response = await axios.post<{ results: SearchResult[] }>(
      `${API_BASE_URL}/tools/search`,
      {
        query,
        max_results: maxResults,
      }
    )
    return response.data.results
  } catch (error) {
    console.error('Web search failed:', error)
    return [] // Return empty array on failure
  }
}

export async function ragQuery(
  query: string,
  fileIds: string[] = [],
  topK = 3
): Promise<RAGResult[]> {
  try {
    const response = await axios.post<{ results: RAGResult[] }>(
      `${API_BASE_URL}/tools/rag/query`,
      {
        query,
        file_ids: fileIds,
        top_k: topK,
      }
    )
    return response.data.results
  } catch (error) {
    console.error('RAG query failed:', error)
    return [] // Return empty array on failure
  }
}
